sway-core/src/asm_generation/fuel/optimizations.rs

```diff
@@ -57,145 +57,184 @@ impl AbstractInstructionSet {
-                    Some(RegContents::Constant(c1)) if c1.checked_add(c2).is_some() => {
-                        reg_contents.insert(dest.clone(), RegContents::Constant(c1 + c2));
-                        record_new_def(latest_version, dest);
-                    }
-                    Some(RegContents::BaseOffset(base_reg, offset))
-                        if get_def_version(latest_version, &base_reg.reg) == base_reg.ver
-                            && offset.checked_add(c2).is_some() =>
-                    {
-                        reg_contents.insert(
-                            dest.clone(),
-                            RegContents::BaseOffset(base_reg.clone(), offset + c2),
-                        );
-                        record_new_def(latest_version, dest);
-                    }
-                    _ => {
-                        let base = VRegDef {
-                            reg: opd1.clone(),
-                            ver: get_def_version(latest_version, opd1),
-                        };
-                        reg_contents.insert(dest.clone(), RegContents::BaseOffset(base, c2));
-                        record_new_def(latest_version, dest);
-                    }
-                }
-            }
             match &mut op.opcode {
-                either::Either::Left(op) => match op {
-                    VirtualOp::ADD(dest, opd1, opd2) => {
-                        // We don't look for the first operand being a constant and the second
-                        // one a base register. Such patterns must be canonicalised prior.
-                        let Some(&RegContents::Constant(c2)) = reg_contents.get(opd2) else {
-                            reg_contents.remove(dest);
-                            record_new_def(&mut latest_version, dest);
-                            continue;
-                        };
-                        process_add(&mut reg_contents, &mut latest_version, dest, opd1, c2);
-                    }
-                    VirtualOp::ADDI(dest, opd1, opd2) => {
-                        let c2 = opd2.value as u64;
-                        process_add(&mut reg_contents, &mut latest_version, dest, opd1, c2);
-                    }
-                    VirtualOp::MUL(dest, opd1, opd2) => {
-                        match (reg_contents.get(opd1), reg_contents.get(opd2)) {
-                            (Some(RegContents::Constant(c1)), Some(RegContents::Constant(c2))) => {
-                                reg_contents.insert(dest.clone(), RegContents::Constant(c1 * c2));
+                either::Either::Left(op) => {
+                    match op {
+                        VirtualOp::ADD(dest, opd1, opd2) => {
+                            // We don't look for the first operand being a constant and the second
+                            // one a base register. Such patterns must be canonicalised prior.
+                            if let Some(&RegContents::Constant(c2)) = reg_contents.get(opd2) {
```

Comment:
Am I correct in understanding that this is a bug fix? (unrelated to MOVs). We're missing an optimization opportunity here

Reply:
This should be the same code, but with a `if let` to accommodate that this runs inside a `retaint_mut` now.

Reply:
ah got it. I read that as an `if let` statement both before and after your change. But it's `let else` before and `if let` now.